libzfs_core/_libzfs_core.py

```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+    return (rfd, wfd)
+
+
+_PIPE_RECORD_FORMAT = 'IBBBB'
+_PIPE_RECORD_SIZE = struct.calcsize(_PIPE_RECORD_FORMAT)
+
+
+@_uncommitted(lzc_list)
+def lzc_get_props(name):
+    '''
+    Get properties of the ZFS dataset.
+
+    :param str name: the name of the dataset.
+    :return: a dictionary mapping the property names to their values.
+    :rtype: dict of str:Any
+
+    .. note::
+        The value of ``clones`` property is a `list` of clone names
+        as byte strings.
+
+    .. warning::
+        The returned dictionary does not contain entries for properties
+        with default values.  One exception is the ``mountpoint`` property
+        for which the default value is derived from the dataset name.
+    '''
+    (fd, other_fd) = lzc_list(name, {'recurse': 1})
+    entry = None
+    try:
+        while True:
+            record_bytes = os.read(fd, _PIPE_RECORD_SIZE)
+            if not record_bytes:
+                break
+            (size, _, err, _, _) =  struct.unpack(_PIPE_RECORD_FORMAT, record_bytes)
+            xlate.lzc_get_props_xlate_error(err, name)
+            if size == 0:
+                break
+            data_bytes = os.read(fd, size)
+            result = {}
+            with nvlist_out(result) as nvp:
+                ret = _lib.nvlist_unpack(data_bytes, size, nvp, 0)
+            if ret != 0:
+                raise ZFSError(ret, "Failed to unpack list data")
+            entry = result['name']
+            if entry == name:
+                break
+            else:
+                entry = None
+    finally:
+        os.close(other_fd)
+        os.close(fd)
+
+    if entry is None:
+        raise exceptions.DatasetNotFound(name)
+    result = result['properties']
+    # In most cases the source of the property is uninteresting and the
+    # value alone is sufficient.  One exception is the 'mountpoint'
+    # property the final value of which is not the same as the inherited
+    # value.
+    mountpoint = result.get('mountpoint')
+    if mountpoint is not None:
+        mountpoint_src = mountpoint['source']
+        mountpoint_val = mountpoint['value']
+        # 'source' is the name of the dataset that has 'mountpoint' set
+        # to a non-default value and from which the current dataset inherits
+        # the property.  'source' can be the current dataset if its
+        # 'mountpoint' is explicitly set.
+        # 'source' can also be a special value like '$recvd', that case
+        # is equivalent to the property being set on the current dataset.
+        if not mountpoint_src.startswith('$'):
+            mountpoint_val = mountpoint_val + name[len(mountpoint_src):]
+    else:
+        mountpoint_val = '/' + name
+    result = { k: v['value'] for k, v in result.iteritems() }
+    if result.has_key('clones'):
+        result['clones'] = [ x for x in result['clones'] ]
+    result['mountpoint'] = mountpoint_val
+    return result
+
+
+@_uncommitted(lzc_list)
+def lzc_list_children(name):
```

Comment:
I think that I have "outsmarted" myself with `is_supported` and `_uncommitted`... a fix is coming.